### PR TITLE
HTS-317: Initial implementation of UNAUTHORIZED.

### PR DESCRIPTION
**Pull request by adwelly**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala [hunk 1]
```diff
@@ -175,10 +176,14 @@ class SquidController @Inject()(val messagesApi: MessagesApi) extends BaseContro
...
   def createAccount(): Action[AnyContent] = Action { request =>
-    val mimeType = request.headers.toSimpleMap.get(CONTENT_TYPE)
+    val headerMap = request.headers.toSimpleMap
+    val mimeType = headerMap.get(CONTENT_TYPE)
+    val authKey = headerMap.get(nsiHeaderKey)
 
     mimeType match {
-      case Some("application/json") => processBody(request)
+      case Some("application/json") =>
+        val headers = request.headers.toMap
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala]: isn't this `headerMap` above?

File: test/uk/gov/hmrc/helptosavestub/controllers/SquidControllerSpec.scala [hunk 2]
```diff
@@ -146,20 +146,26 @@ class SquidControllerSpec extends UnitSpec with WithFakeApplication {
     Json.toJson(goodAccount copy (contactDetails = cd))
   }
 
-  private val fakeRequest = FakeRequest().withJsonBody(Json.toJson(goodAccount)).withHeaders((CONTENT_TYPE, "application/json"))
+  private val fakeRequest = FakeRequest().withJsonBody(Json.toJson(goodAccount)).withHeaders((CONTENT_TYPE, "application/json"), "Authorization1" -> "someValue")
```
Comment (AndyHWChung) [inline test/uk/gov/hmrc/helptosavestub/controllers/SquidControllerSpec.scala]: These tests will fail if we ever change the authorisation header key in the `application.conf` - it's safer to read this from the config here if that's what's happening in the controller

File: app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala [hunk 3]
```diff
@@ -175,10 +176,14 @@ class SquidController @Inject()(val messagesApi: MessagesApi) extends BaseContro
...
+        if (authKey.isDefined) processBody(request) else Unauthorized
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala]: Do we want to check the username and password?

File: app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala [hunk 4]
```diff
@@ -175,10 +176,12 @@ class SquidController @Inject()(val messagesApi: MessagesApi) extends BaseContro
   }
 
   def createAccount(): Action[AnyContent] = Action { request =>
-    val mimeType = request.headers.toSimpleMap.get(CONTENT_TYPE)
+    val headerMap = request.headers.toSimpleMap
+    val mimeType = headerMap.get(CONTENT_TYPE)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala]: are these used anywhere ?

Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala]: there's a pattern match on `mimetype` below and `authKey` uses `headerMap`

File: app/uk/gov/hmrc/helptosavestub/config/microserviceGlobal.scala [hunk 5]
```diff
@@ -29,6 +29,7 @@ import uk.gov.hmrc.play.filters.MicroserviceFilterSupport
 
 object ControllerConfiguration extends ControllerConfig {
   lazy val controllerConfigs = Play.current.configuration.underlying.as[Config]("controllers")
+  val nsiHeaderKey = Play.current.configuration.underlying.as[Config]("nsi").getString("authorization.header-key")
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavestub/config/microserviceGlobal.scala]: Could also have written this as:
```scala
val nsiHeaderKey = Play.current.configuration.underlying.getString("nsi.authorization.header-key")
```

File: app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala (hunk 4 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavestub/controllers/SquidController.scala]: Yes. They are used to check that the mime type and authorization code is present in the header.

File: app/uk/gov/hmrc/helptosavestub/config/microserviceGlobal.scala (hunk 5 above)
Comment (adwelly) [inline app/uk/gov/hmrc/helptosavestub/config/microserviceGlobal.scala]: Looks better, I agree

**Pull request merged**